infrastructure/infrastructure.py

```diff
@@ -0,0 +1,638 @@
+    except ClientError as e:
+        error_code = e.response["Error"]["Code"]
+        already_attached = "already attached" in str(e).lower()
+        if already_attached or error_code == "InvalidInputException":
+            logger.warning(
+                "Static IP %s already attached to %s, continuing...",
+                ip_name,
+                instance.name,
+            )
+        else:
+            raise
+
+    instance.public_ip = ls.get_static_ip(staticIpName=ip_name)["staticIp"]["ipAddress"]
+    return instance
+
+
+def run_setup_script(instance: Instance) -> Instance:
+    tailscale_auth_key = os.getenv("TAILSCALE_AUTH_KEY")
+
+    max_retries = 20
+    retry_delay = 15
+
+    for attempt in range(max_retries):
+        try:
+            with Connection(
+                host=instance.public_ip,
+                user="ubuntu",
+                connect_kwargs={
+                    "key_filename": str(PRIVATE_KEY_PATH),
+                    "banner_timeout": 30,
+                    "auth_timeout": 30,
+                },
+                connect_timeout=10,
+            ) as ssh:
+                logger.info("Connected to %s via SSH", instance.name)
+                break
+        except Exception:
+            if attempt < max_retries - 1:
+                logger.info(
+                    "SSH connection to %s failed (attempt %d/%d), retrying in %ds...",
+                    instance.name,
+                    attempt + 1,
+                    max_retries,
+                    retry_delay,
+                )
+                time.sleep(retry_delay)
+            else:
+                logger.exception(
+                    "Failed to connect to %s after %d attempts",
+                    instance.name,
+                    max_retries,
+                )
+                raise
+
+    with Connection(
+        host=instance.public_ip,
+        user="ubuntu",
+        connect_kwargs={
+            "key_filename": str(PRIVATE_KEY_PATH),
+            "banner_timeout": 30,
+            "auth_timeout": 30,
+        },
+        connect_timeout=10,
+    ) as ssh:
+        ssh.run("sudo apt-get update && sudo apt-get -y upgrade")
+
+        ssh.run(
+            "curl -fsSL https://get.docker.com -o get-docker.sh && sh get-docker.sh"
+        )
+        ssh.run("sudo usermod -aG docker $USER")
+        ssh.run("sudo systemctl enable docker && sudo systemctl restart docker")
+
+        ssh.run("sudo mkdir -p /etc/docker")
+        ssh.run(
+            "echo '{"
+            '"swarm-default-advertise-addr": "tailscale0", '
+            '"mtu": 1280'
+            "}' | sudo tee /etc/docker/daemon.json"
+        )
+        ssh.run("sudo systemctl restart docker")
+
+        ssh.run("curl -fsSL https://tailscale.com/install.sh | sh")
+        ssh.run(
+            f"sudo tailscale up --auth-key={tailscale_auth_key} "
+            f"--hostname={instance.name} --ssh --accept-routes"
```

Comment:
There's definitely brittleness in here with all of the scripting being called but I'm sure it's fine for now.